rust/sbd-client/src/lib.rs

```diff
@@ -0,0 +1,320 @@
+
+    /// Setting this to `true` disables certificate verification on `wss://`
+    /// scheme. WARNING: this is a dangerous configuration and should not
+    /// be used outside of testing (i.e. self-signed tls certificates).
+    pub danger_disable_certificate_check: bool,
+}
+
+impl Default for SbdClientConfig {
+    fn default() -> Self {
+        Self {
+            out_buffer_size: MAX_MSG_SIZE * 8,
+            allow_plain_text: false,
+            danger_disable_certificate_check: false,
+        }
+    }
+}
+
+/// SbdClient represents a single connection to a single sbd server
+/// through which we can communicate with any number of peers on that server.
+pub struct SbdClient {
+    send_buf: Arc<tokio::sync::Mutex<send_buf::SendBuf>>,
+    read_task: tokio::task::JoinHandle<()>,
+    write_task: tokio::task::JoinHandle<()>,
+}
+
+impl Drop for SbdClient {
+    fn drop(&mut self) {
+        self.read_task.abort();
+        self.write_task.abort();
+    }
+}
+
+impl SbdClient {
+    /// Connect to the remote sbd server.
+    pub async fn connect<C: Crypto>(
+        url: &str,
+        crypto: &C,
+    ) -> Result<(Self, String, PubKey, MsgRecv)> {
+        Self::connect_config(url, crypto, SbdClientConfig::default()).await
+    }
+
+    /// Connect to the remote sbd server.
+    pub async fn connect_config<C: Crypto>(
+        url: &str,
+        crypto: &C,
+        config: SbdClientConfig,
+    ) -> Result<(Self, String, PubKey, MsgRecv)> {
+        use base64::Engine;
+        let full_url = format!(
+            "{url}/{}",
+            base64::engine::general_purpose::URL_SAFE_NO_PAD
+                .encode(crypto.pub_key())
+        );
+
+        let (mut send, mut recv) = raw_client::WsRawConnect {
+            full_url: full_url.clone(),
+            max_message_size: MAX_MSG_SIZE,
+            allow_plain_text: config.allow_plain_text,
+            danger_disable_certificate_check: config
+                .danger_disable_certificate_check,
+        }
+        .connect()
+        .await?;
+
+        let raw_client::Handshake {
+            limit_byte_nanos,
+            limit_idle_millis,
+            bytes_sent,
+        } = raw_client::Handshake::handshake(&mut send, &mut recv, crypto)
+            .await?;
+
+        let send_buf = send_buf::SendBuf::new(
+            send,
+            config.out_buffer_size,
+            (limit_byte_nanos as f64 * 1.1) as u64,
+            std::time::Duration::from_millis((limit_idle_millis / 2) as u64),
+            bytes_sent,
+        );
+        let send_buf = Arc::new(tokio::sync::Mutex::new(send_buf));
+
+        let send_buf2 = send_buf.clone();
+        let (recv_send, recv_recv) = tokio::sync::mpsc::channel(4);
+        let read_task = tokio::task::spawn(async move {
+            while let Ok(data) = recv.recv().await {
+                let data = Msg(data);
+
+                match match data.parse() {
+                    Ok(data) => data,
+                    Err(_) => break,
```

Comment:
Yep, when it breaks out of the loop it closes the connection, (through the `send_buf2.lock().await.close().await` call below).

There really aren't a lot of varied error cases... If you were able to establish the connection, but now it no longer exists, all you know is that the connection was closed. By design, you don't know it was closed due to rate-limiting, for example (because we don't want to waste system resources sending data to potential denial of service node).